components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2433,10 +2539,38 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         let mut meta = self.ctx.store_meta.lock().unwrap();
         debug!(
             "check snapshot range";
-            "region_id" => self.region_id(),
+            "region_id" => self.fsm.region_id(),
             "peer_id" => self.fsm.peer_id(),
             "prev_region" => ?prev_region,
         );
+
+        // Remove this region's snapshot region from the `pending_snapshot_regions`
+        // The `pending_snapshot_regions` is only used to occupy the key range, so if this
+        // peer is added to `region_ranges`, it can be remove from `pending_snapshot_regions`
+        meta.pending_snapshot_regions
+            .retain(|r| self.fsm.region_id() != r.get_id());
+
+        // Remove its source peers' metadata
+        for r in &apply_result.destroyed_regions {
```

Comment:
How are read delegates removed in the old implementation? Is it a bug?

Reply:
Like the normal merge process, for atomicity, we delete some source region's metadata and change some target region's metadata in the target peer function `on_ready_commit_merge`. Then it sends destroy msg to the source peer.